internal/controllers/rollout/synthesizer.go

```diff
@@ -0,0 +1,127 @@
+	}
+	logger = logger.WithValues("synthesizerName", syn.Name, "synthesizerNamespace", syn.Namespace, "synthesizerGeneration", syn.Generation)
+
+	compList := &apiv1.CompositionList{}
+	err = c.client.List(ctx, compList, client.MatchingFields{
+		manager.IdxCompositionsBySynthesizer: syn.Name,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing compositions: %w", err)
+	}
+
+	// randomize list to avoid always rolling out changes in the same order
+	rand.Shuffle(len(compList.Items), func(i, j int) { compList.Items[i] = compList.Items[j] })
+
+	for _, comp := range compList.Items {
+		comp := comp
+		logger := logger.WithValues("compositionName", comp.Name, "compositionNamespace", comp.Namespace, "compositionGeneration", comp.Generation)
+
+		// Compositions aren't eligible to receive an updated synthesizer when:
+		// - They haven't ever been synthesized (they'll use the latest inputs anyway)
+		// - They are currently being synthesized or deleted
+		// - They are already pending resynthesis
+		// - They are already in sync with the latest inputs
+		if comp.Status.CurrentSynthesis == nil || comp.Status.CurrentSynthesis.Synthesized == nil || comp.DeletionTimestamp != nil || comp.Status.PendingResynthesis != nil || isInSync(&comp, syn) {
+			continue
+		}
+
+		comp.Status.PendingResynthesis = ptr.To(metav1.Now())
+		err = c.client.Status().Update(ctx, &comp)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("swapping compisition state: %w", err)
+		}
+
+		logger.V(1).Info("staged resynthesis of composition because its synthesizer changed")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func isInSync(comp *apiv1.Composition, syn *apiv1.Synthesizer) bool {
+	return comp.Status.CurrentSynthesis.ObservedSynthesizerGeneration >= syn.Generation
+}
+
+func newCompositionHandler() handler.EventHandler {
+	return &handler.Funcs{
+		CreateFunc: func(ctx context.Context, ce event.CreateEvent, rli workqueue.RateLimitingInterface) {
+			// No need to handle creation events since the status will always be nil.
+		},
+		DeleteFunc: func(ctx context.Context, de event.DeleteEvent, rli workqueue.RateLimitingInterface) {
+			// We don't handle deletes on purpose, since a composition being deleted can only ever
+			// result in the cooldown period being shortened i.e. we lose track of a more recent
+			// rollout event.
+			//
+			// It's okay that this state can be lost, since it falls within the promised semantics
+			// of this controller. But ideally we can avoid it when possible.
+		},
+		UpdateFunc: func(ctx context.Context, ue event.UpdateEvent, rli workqueue.RateLimitingInterface) {
+			newComp, ok := ue.ObjectNew.(*apiv1.Composition)
+			if !ok {
+				logr.FromContextOrDiscard(ctx).V(0).Info("unexpected type given to newCompositionToSynthesizerHandler")
+				return
+			}
+
+			oldComp, ok := ue.ObjectOld.(*apiv1.Composition)
+			if !ok {
```

Comment:
Can this ever happen?

Reply:
The previous resource isn't sent from the server in watch events, so (if memory serves 😬 ) it's populated with the last version in the informer cache when the new version is received. So it can be nil if the process has crashed.